### ISSUE TITLE
Deflake Kubenet e2e suite

### ISSUE THREAD

**Issue by freehan** [kubernetes/kubernetes#24211]
I gathered e2e test results of `kubernetes-e2e-gce-kubenet` suite for last 100 runs. The following is the list of flaky tests. Some might not be related to `kubenet`. I will start with the service tests. Any help would be appreciated. 

```
DNS should provide DNS for pods for Hostname and Subdomain Annotation
DNS should provide DNS for services [Conformance]
DNS should provide DNS for the cluster [Conformance]
Deployment RecreateDeployment should delete old pods and create new ones
Deployment RollingUpdateDeployment should delete old pods and create new ones
Dynamic provisioning [k8s.io] DynamicProvisioner should create and delete persistent volumes
Horizontal pod autoscaling (scale resource: CPU) [k8s.io] ReplicationController light Should scale from 1 pod to 2 pods
Horizontal pod autoscaling (scale resource: CPU) [k8s.io] ReplicationController light Should scale from 2 pods to 1 pod using HPA version v1
Kubectl client [k8s.io] Guestbook application should create and stop a working application [Conformance]
Kubectl client [k8s.io] Kubectl expose should create services for rc [Conformance]
Kubectl client [k8s.io] Simple pod should support exec through an HTTP proxy
Kubectl client [k8s.io] Update Demo should create and stop a replication controller [Conformance]
Kubectl client [k8s.io] Update Demo should do a rolling update of a replication controller [Conformance]
Kubectl client [k8s.io] Update Demo should scale a replication controller [Conformance]
Monitoring should verify monitoring pods and all cluster nodes are available on influxdb using heapster.
Networking should provide Internet connection for containers [Conformance]
Pods should *not* be restarted with a /healthz http liveness probe [Conformance]
PreStop should call prestop when killing a pod [Conformance]
Proxy version v1 should proxy through a service and a pod [Conformance]
ReplicaSet should serve a basic image on each replica with a private image
ReplicaSet should serve a basic image on each replica with a public image [Conformance]
ReplicationController should serve a basic image on each replica with a public image [Conformance]
Services should be able to create a functioning NodePort service
Services should be able to up and down services
Services should serve multiport endpoints from pods [Conformance]
```

Detail list: https://docs.google.com/spreadsheets/d/1Le8UfN_xj7FzI-NuVZ1le-_a8IW62gM40SAWEWv8Xhs/edit#gid=0
Jenkins project:  http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-kubenet/

**Comment by dchen1107** [comment]
cc/ @jlowdermilk Cluster team is working on to make kubenet default. Once we switch this to production, we don't need to restart docker to configure cbr0.

**Comment by timothysc** [comment]
just hit "RecreateDeployment should delete old pods and create new ones"

**Comment by freehan** [comment]
After a bunch PRs got merged, the kubenet suite is much more stable. One more known flake can be fixed by #23711

**Comment by freehan** [comment]
Just bump into a failure: https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-kubenet/1825/

One of the nodes (`jenkins-e2e-minion-7o0s`) got panic: 
```
I0511 11:38:52.400551   16431 kubenet_linux.go:310] TearDownPod took 19.153595ms for e2e-tests-clientset-e3982/pod8f3df856-176c-11e6-b52f-0242ac110015
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x40 pc=0x512932]

goroutine 558 [running]:
panic(0x1eba080, 0xc82000c090)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/kubelet/network/kubenet.(*kubenetNetworkPlugin).TearDownPod(0xc820117f10, 0xc820d10cf8, 0x19, 0xc820d10cd0, 0x27, 0x22653f8, 0x6, 0xc82027b6c0, 0x40, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/network/kubenet/kubenet_linux.go:336 +0x842
k8s.io/kubernetes/pkg/kubelet/dockertools.(*DockerManager).killPodWithSyncResult(0xc8200e28c0, 0x0, 0xc82072ea74, 0x24, 0xc820d10cd0, 0x27, 0xc820d10cf8, 0x19, 0xc8200c6580, 0x1, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockertools/manager.go:1241 +0x11e5
k8s.io/kubernetes/pkg/kubelet/dockertools.(*DockerManager).KillPod(0xc8200e28c0, 0x0, 0xc82072ea74, 0x24, 0xc820d10cd0, 0x27, 0xc820d10cf8, 0x19, 0xc8200c6580, 0x1, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockertools/manager.go:1174 +0x70
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).killPod(0xc820214900, 0x0, 0xc820d10d20, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1651 +0xda
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).podKiller.func1(0xc820214900, 0x0, 0xc820d10d20, 0xc820230a80)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2269 +0x19c
created by k8s.io/kubernetes/pkg/kubelet.(*Kubelet).podKiller
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2273 +0x2aa
```

Looks like it crashed at shaper reset.: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/network/kubenet/kubenet_linux.go#L336


@dcbw Can you take a look?
